### PR TITLE
Add missing features

### PR DESCRIPTION
**Pull request by hicom150**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix
[x] New feature
[ ] Other, please explain:

**What changes did you make? (Give an overview)**
I've added the missing features one by commit

- Add better unit test (I've also added few integration tests but they are skipped)
- Add IPv6 support
- Add flags support
- Add message rate check

**Which issue (if any) does this pull request address?**

**Is there anything you'd like reviewers to focus on?**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -100,23 +127,64 @@ module.exports = () => {
...
+          compact2string.multi6(message.dropped6).forEach(peer => {
+            delete this._remoteAddedPeers[peer]
+            if (!(peer in this._remoteDroppedPeers)) {
+              this._remoteDroppedPeers[peer] = { ip: 6 }
+              this.emit('dropped', peer)
+            }
+          })
+        } catch (err) {
+          // drop invalid messages
+        }
```
Comment (KayleePop) [inline index.js:187]: seems like these could be more DRY

**Review by KayleePop — APPROVED** [review]
didn't do a super thorough review, so please check over everything once more before merging, but LGTM

File: README.md [hunk 2]
```diff
@@ -112,28 +114,50 @@ wire.ut_pex.reset()
...
+wire.ut_pex.addPeer(peer, flags)
+```
+
+### addPeer6
+
+Adds an IPv6 peer to the locally discovered peer list to send with the next PEX message.
+
+```js
+const peer = '[::1]:6889'
+const flags = 0x06
```
Comment (KayleePop) [inline README.md:132]: should we make an API for the accepted flags? Like we could use function arguments like `supports_utp` `prefers_encryption` etc and then set them as properties on the peer object like `peer.supports_utp === true`

we should decide now before cementing an API

Comment (KayleePop) [reply README.md:132]: actually, not sure if I fully understand the BEP, but flags should be a list not just one number I think?

File: index.js [hunk 3]
```diff
@@ -125,16 +193,42 @@ module.exports = () => {
...
-
-      var addedFlags = Buffer.concat(localAdded.map(() => {
-        // TODO: support flags
-        return Buffer.from([0])
-      }))
+      const localAdded = Object.keys(this._localAddedPeers).slice(0, PEX_MAX_PEERS)
+      const localDropped = Object.keys(this._localDroppedPeers).slice(0, PEX_MAX_PEERS)
+
+      const _isIPv4 = (peers, addr) => peers[addr].ip === 4
+      const _isIPv6 = (peers, addr) => peers[addr].ip === 6
```
Comment (KayleePop) [inline index.js:200]: maybe this would be simpler if there were a separate `_localAddedPeers` for ipv6

File: README.md (hunk 2 above)
Comment (hicom150) [reply README.md:132]: Yeah! I think it is a good idea handle the flags as properties 👍 I'll try to work on it a bit 😉

File: index.js (hunk 3 above)
Comment (hicom150) [reply index.js:200]: The good part of having it in the same object is that when we have to choose only 50 peers, there are in the same "bucket" so you don't have to be taking them from different places 😅
Giving that the filtering seems quite cheep, I seen more "pros" than "cons". What do you think?

**Comment by hicom150** [comment]
As always, great feedback @KayleePop 👍
I'll try to work in flags, I'll let you know when It's ready 😉

File: index.js [hunk 4]
```diff
@@ -125,16 +193,42 @@ module.exports = () => {
...
+
+      const dropped6 = localDropped
+        .filter(k => _isIPv6(this._localDroppedPeers, k))
+        .map(string2compact)
+        .reduce((acc, cur) => Buffer.concat([acc, cur]), Buffer.alloc(0))
+
+      const addedFlags = localAdded
+        .filter(k => _isIPv4(this._localAddedPeers, k))
+        .map(k => _flags(this._localAddedPeers, k))
+        .reduce((acc, cur) => Buffer.from([cur]), Buffer.alloc(0))
```
Comment (KayleePop) [inline index.js:226]: should this be 

```js
.map(flagInt => Buffer.from([flagInt]))
.reduce((acc, cur) => Buffer.concat([acc, cur]), Buffer.alloc(0))
```

? (and for `add6Flags` too)

File: index.js (hunk 3 above)
Comment (KayleePop) [reply index.js:200]: hm, yea

It might be more readable if you split the localAdded and localDropped into ipv4 and ipv6 local variables though, instead of filtering multiple times below. Not sure. It's just kind of hard to parse

File: index.js (hunk 1 above)
Comment (hicom150) [reply index.js:176]: Addressed that in the last refactor 👍

File: README.md (hunk 2 above)
Comment (hicom150) [reply README.md]: I've added an API for flags, hope it is more clear now 😉

File: index.js (hunk 3 above)
Comment (hicom150) [reply index.js:219]: I've refactored this part, hope it is more clear now 😅

File: index.js (hunk 4 above)
Comment (hicom150) [reply index.js]: Nice catch! This was actually a bug, so I fixed and added more test 👍

**Comment by hicom150** [comment]
@KayleePop I've added two commits **Add better flags API** and **Refactor with review suggestions** to address your suggestions.
I'd appreciate if you could give it a last review just to check if everything fits 😅

Thank you! ❤️

**Comment by KayleePop** [comment]
great! looks good I think

**Comment by feross** [comment]
Nice work, @hicom150. This is really great. We finally addressed those TODOs which have been in the readme since the very earliest days!

The only last bit of feedback I have is that it might be nice to camelCase the properties in the flags object.
